### DISCUSSION TITLE
How multiple source sets can be managed in Maven

### DISCUSSION THREAD

**Discussion by hendrikebbers** [support-and-care/maven-support-and-care#37] [Ideas for Funding]
This is based on the discussion in https://github.com/OpenElements/maven-support-care/discussions/11

An idea (copied from https://github.com/OpenElements/maven-support-care/discussions/11) might look like this:

```
<sources>
  <source module="org.foo.bar" release="11">src/org.foo.bar/java</source>
  <source module="org.foo.biz" release="11">src/org.foo.biz/java</source>
  <source module="org.foo.biz" release="17">src/org.foo.biz/java17</source>
<sources>
```

Goal of this milestone should not be to have a final 100% solid integration in Maven but to have a definition for the future with a working prototype. Next to that parts of the prototype that does not make Maven 4 incompatible can be merged directly to Maven.

**Comment by rmannibucau** [comment]
Why not using the built in support of compiler (+resources if needed) plugin thru executions? 🤔

**Reply by desruisseaux** [reply]
This information is not specific to the compiler plugin. It will also be needed by the Javadoc plugin, and potentially all plugins dealing with source code. Another reason is because in the current POM model, the source directories are already specified in the `<build>` section. I think that adding different places where to declare sources would add confusion. Foe example, does the compiler plugin configuration override or add to the `<build>` configuration?

Currently, the source directories are declared in the following locations (from [Maven POM reference](https://maven.apache.org/pom.html#the-super-pom)):

```xml
<build>
  <sourceDirectory>${project.basedir}/src/main/java</sourceDirectory>
  <scriptSourceDirectory>${project.basedir}/src/main/scripts</scriptSourceDirectory>
  <testSourceDirectory>${project.basedir}/src/test/java</testSourceDirectory>
  <resources>
    <resource>
      <directory>${project.basedir}/src/main/resources</directory>
    </resource>
  </resources>
</build>
```

The proposal is to modify the above model. It does not fundamentally changes the model, but makes the above both more flexible and more consistent. The above could be replaced by the following (note how the resources become handled like any other sources). Note that it would be legal to have more than one directory of the same type (e.g. many resource directories):

```xml
<build>
  <sources>
    <directory type="main">${project.basedir}/src/main/java</directory>
    <directory type="script">${project.basedir}/src/main/scripts</directory>
    <directory type="test">${project.basedir}/src/test/java</directory>
    <directory type="resource">${project.basedir}/src/main/resources</directory>
  </sources>
</build>
```

The above proposal would allow, for example, to add `<directory type="snippet">` (new in Java 18) without the need to change the POM model. More important for the JPMS support project, it would allow to add `module` attribute. The `release` attribute could also be opportunisically added.

**Reply by rmannibucau** [reply]
the default model assumes you use maven conventions, what you propose makes it completely generic - as the type change - which means:

1. the conventions are pointless
2. IDE will have a hard time to integrate it since "type" does not mean anything anymore by default
3. it does not enable anything since this is already possible but the specific parts belongs to plugin configurations

So not sure it would be that good and it is clearly changing the maven paradigm - always got a hard no on the list when trying to push such a change for other topics like frontend build - so guess this is something to discuss and make validated on the list before discussing cause it can likely never happen.

**Reply by desruisseaux** [reply]
@cstamas would know better than me, but my understanding is that the conventions are defined by the super POM, which is the link that I provided in above comment. If the super POM was modified, conventions would follow. Note however that the proposal that I wrote in above comment are the same conventions, only written differently. Finally, I don't think that this new format is more difficult than the current one for IDE. It it the same thing, with only an element name replaced by an attribute value. It would require changes in IDE, but I don't see why it would be more difficult than the current model.

The proposal enables new features: there is no way to add a "snippet" type of sources in current POM without changing the XML schema, contrarily to the proposal. A "snippet" type of sources applies not only to the Javadoc plugin, but also to the compiler plugin if we want to ensure that the snippets are valid code. Generally speaking, sources apply to the project as a whole: they are information shared by many plugins (e.g. main sources apply to compiler, javadoc, checkstyle…), they are not specific to a single plugin.

Finally, the above proposal just provides the same functionality written in a different way. But the real goal is to allow the addition of `module` and `release` attributes. The Java module name, for example, is needed for [this Java compiler API](https://docs.oracle.com/en/java/javase/22/docs/api/java.compiler/javax/tools/StandardJavaFileManager.html#setLocationForModule(javax.tools.JavaFileManager.Location,java.lang.String,java.util.Collection)). It can't be done with current model. Again, a configuration specific to the compiler plugin would not allow to reuse that information e.g. for Javadoc.

**Reply by rmannibucau** [reply]
>  conventions are defined by the super POM

nop, also by the design and I think maven sticks to 1 target per module (ie one module per java target and one aggregator project - not saying it is good just saying a "one step" solution which does not break the whole convention is bad, either you change the core paradigm - and it is not a directory.type you need but to drop all the resources definition to host them 100% in plugins, or you don't change it, anything in between looks wrong to me)

> The proposal enables new features: there is no way to add a "snippet" type of sources in current POM without changing the XML

yes, was the intent but maven already enables this use case thanks plugin by design, you just look for another build tool (so maybe move from https://github.com/OpenElements/maven-support-care/discussions/34 short term instead of patching maven by small point to build a tool which has no more link with maven but starts to over-generic)

> Finally, the above proposal just provides the same functionality written in a different way.

I'm always in the camp to keep backward compat so if there is no new feature maybe just avoid it.
You will note that release attribute can *not* be a child of build but must stick to compiler configuration (or javadoc etc - being said they don't have and often are not aligned for libraries).
Module is likely at the same level.

From my window you have a simple extreme use case and you want to drop duplicate configuration for your use case but it looks like it creates noise for common use case and other (libraries targetting using multiple jvm) cases.

Once again it got discussed multiple times on maven list about having mutiple source sets and always got rejected to get a custom lifecycle/packaging outside maven, I don't see yet why JPMS or mjar would be different (and it is not hard at all today already).

> The Java module name, for example, is needed for [this Java compiler API](https://docs.oracle.com/en/java/javase/22/docs/api/java.compiler/javax/tools/StandardJavaFileManager.html#setLocationForModule(javax.tools.JavaFileManager.Location,java.lang.String,java.util.Collection)). It can't be done with current model.

Why? It is just one more configuration in the plugin.

**Reply by cstamas** [reply]
> nop, also by the design and I think maven sticks to 1 target per module (ie one module per java target and one aggregator project - not saying it is good just saying a "one step" solution which does not break the whole convention is bad, either you change the core paradigm - and it is not a directory.type you need but to drop all the resources definition to host them 100% in plugins, or you don't change it, anything in between looks wrong to me)

That remains, "1 module 1 artifact" is what I think you wanted to say, but there are multi-release JARs these days, and many other things that alter the meaning of that above: it is _not more_ one compiler release per module for example as it was in Maven3.

@gnodet sorry, wanted to ping Guillaume

**Reply by desruisseaux** [reply]
> If there is no new feature maybe just avoid it.

The example that I provided in above comment has no new features because I wanted to show how the proposal can map to exactly what currently exists in Maven 3. But it does enable new features, and that is the goal. The example below is not even about JPMS:

```xml
<build>
  <sources>
    <directory type="main" release="8">${project.basedir}/src/main/java</directory>
    <directory type="main" release="17">${project.basedir}/src/main/java17</directory>
    <directory type="test">${project.basedir}/src/test/java</directory>
  </sources>
</build>
```

Multi-release JAR files did not existed when Maven 3 was designed. It is not impossible to do multi-releases with Maven 3, but it is non-obvious. Now that the Java language has evolved, Maven should evolve too since leveraging the new Java features with the Maven 3 model become convolved.

> Why? It is just one more configuration in the plugin.

I already replied to that: because the same configuration needs to be shared by many plugins. The `javax.tools` API that I pointed to is shared by the compiler and the Javadoc tool. Do you want to force users to copy-and-paste the same block of configuration in compiler and javadoc plugins? (to name just the two most obvious).

**Reply by rmannibucau** [reply]
Foes not enable any feature as explained, belongs to conventions+plugin conf *today*.
Mjar releases have a challenge: it breaks target bytecode (ex: a j11 jar will contain j21 code 1nd some libs are still broken about it) do agree we should ease the usage by a convention but dont think we should make the conf so generic, will create issues and make maven less friendly as everytime we did such a change.

Now it is just my 2cts but last changes in maven breaks its low entrycost 1nd I find it unwelcoming and negative overall for features very few people will care about in practise.

**Reply by desruisseaux** [reply]
Duplicating the same information in many places is entropy. The claim that few peoples care in practice is not backed by facts such as the very existence of this "Support & Care for Apache Maven" project and the fact that they have chosen JPMS as a priority. Users requests for JPMS support exist since at least 2018 (e.g. [MNG-6390](https://issues.apache.org/jira/browse/MNG-6390) — even if that particular one has been closed as "not a problem").

**Reply by rmannibucau** [reply]
Can you detail your entropy comoutation? You assume it will behave as today but in practose it will not. Common example is you dont want javadoc to use j17 sources when you build with j11 as baseline etc so it is NOT comparable to handle mjars and standard/current mode.
Entropy does not increase by using plugin conf.

Facts are just the low requests compared to some other features, this bugtracker is a bad reference IMO cause does not represent list activity nor users feedbacks.
Mjar and jpms always got solved in ways where users were happy (mass statement if you review slack or forums for ex) so yes this is far to be a big thing and even inside maven community mjar usage is not a shared and ack as good thing.

**Reply by cstamas** [reply]
> Mjar and jpms always got solved in ways where users were happy (mass statement if you review slack or forums for ex) so yes this is far to be a big thing and even inside maven community mjar usage is not a shared and ack as good thing.

This sentence is simply untrue: just look at "modern bits" or Maven ecosystem, like latest new projects (mvnd), or latest resolver or latest indexer, they all use MJAR feature.

**Reply by rmannibucau** [reply]
Does not mean it is wrong, just that it passed either under radars (or lack of interest case for these modules) or in force like resolver...

All that must be discussed on the list and hopefully there will be a consensus but take care that this bugtracker opens tickets on topics where there is not yet any agreement (mainly by lack of exchange).
Now if you think you are numerous enough to pass in force you can ignore that too, I dislike it as an asf guy but ill also not fight and can just leave if it goes that way to avoid more friction than we already have.

**Reply by cstamas** [reply]
> Facts are just the low requests compared to some other features, this bugtracker is a bad reference IMO cause does not represent list activity nor users feedbacks.

If I am about to agree with you, that Martin is "biased" and him pointing at JIRA issues is "not representational", I have to state at the same time that your point is "not representational" either: you are referencing to a code base that had literally little if no changes (sans "update deps" I do not count) happened since 15 years, and it was pretty much in state "as it was done" (before or around 15 years ago).

**Comment by desruisseaux** [comment]
My goal with the reference to MNG-6390 was not to said that requests are numerous (I have no data for that), but that they are not recent.

**Reply by rmannibucau** [reply]
Yep but it is like jsr223, when it got created we got a few requests, then nothing for years and recently when a "core" pushed enough it became a maven plugin.
No issue to do the same with mjar (in compiler) or jpms but shouldnt impact the core model yet IMHO (to recheck for v5 if it changed is what I mean).

**Comment by jjohannes** [comment]
Because this was compared to what Gradle offers in earlier comments, I would like to share some info on that. Maybe it helps to put the scope here into perspective. In Gradle these things are called **Source Set** and **Source Directory Set** (small but important destiction!). Both **Source Set** and **Source Directory Set** can be dynamically added to a project. In fact, these are not core concepts but only concepts added by the `java` plugins that are implemented on top of the dynamic "task graph" and "dependency management" systems of Gradle core. I also have [a video and further doc links on the topic](https://github.com/jjohannes/understanding-gradle/tree/main/16_Source_Sets).

**Example**
```
└── src
    ├── main           <- Source Set
    │   ├── java          <- Source Directory Set (it's called Set because in can have several folders, but in practice it usually has one)
    │   ├── kotlin        <- Source Directory Set
    │   └── java21        <- Source Directory Set
    ├── testFixtures   <- Source Set
    │   └── java          <- Source Directory Set
    ├── integrationTest <- Source Set
    │   └── java          <- Source Directory Set
    └── test           <- Source Set
        ├── java          <- Source Directory Set
        └── groovy        <- Source Directory Set
```

- A **Source Set** has a separate context. It may be completely disconnected from other source sets, almost as if it was a separat (sub)project. It has it's own dependency scopes. E.g. if you add the `testFixtures` source set, you have `testFixturesImplementation` and `testFixturesApi` scopes for declaring dependencies. And  there is a `testFixturesCompileClasspath` and a `testFixturesRuntimeClasspath`. So each source set has its own dependencies.
- All **Source Directory Sets** that belong to a source set share the same context but usually have a separate task operating on each of them. E.g. `main/java` is processed by a Java compile task and `main/kotlin` by a Kotlin compile task. But both tasks get the same `compileClasspath` as input.

**In Maven...**
....I would say there are already plugins which behave similar to the **Source Directory Set** concept. For example the `kotlin-maven-plugin`. I think this is quite comparable because they also work in the same context (lifecycle) with the same dependency scopes. Maybe this concept can be generalized so that it becomes easier to add new **Source Directory Sets**. One use case is to have several Java source folders for different Java versions where the compile goal would be executed for each of them with different compiler options. 
For the concept of **Source Set** I cannot really imagine how it would work in Maven. As you have the fixed `compile` and `test-compile` lifecycle phases and the fixed dependency scopes, it does not really map. You could think about that you can define a "source set" and then map it to one of the existing phases and dependency scopes, but then there would be conceptually not much difference to a **Source Directory Set**  I think.

**Reply by cstamas** [reply]
@gnodet ping, and this is what i wanted to achieve with (different kind of) "scopes" and "project paths", if you remember

**Reply by rmannibucau** [reply]
Hmm, I still fail to see how you make it work, do you move the full compiler+javadoc+jlink+... configurations to `source` tag? If not we just end with `<source>src/org.foo.biz/java17</source>` which is what we kind of already have else does it mean we drop default plugin configuration to move it elsewhere (would be very negative since we just forced to define the plugin and therefore the configuration is not an issue)?

The other issue in the original snippet is:

```
  <source module="org.foo.bar" release="11">src/org.foo.bar/java</source>
  <source module="org.foo.biz" release="11">src/org.foo.biz/java</source>
```

Maven community - I exclude myself there ;) - always rejected this idea to build 2 (GA/not related) jars in a single module and make 1 module = 1 multimodule project, so it sounds very fishy and build a mjar is not that hard and would still require more setup anyway (I don't think it would be auto enabled just cause there are 2 sources).

**Comment by desruisseaux** [comment]
The Gradle's concept of "source set" may map approximately (but not completely) to the `type` attribute in above proposal. I think that both approaches would be equivalent for JPMS. The `type` approach may be easier to map to the existing Maven design.

About building many JAR files in a single Maven module, even if the community rejected that proposal in the past, it may be time to reconsider. Because the Java language has evolved. In a Modular Source Hierarchy, those JAR files are not unrelated. Their content are the result of a single call to `javac`, it is not one separated compilation per JAR file. So before to reject this idea, there is two questions to answer:

* Can you compile [this simple test case](https://github.com/Geomatys/MavenModulepathBug/tree/cyclic-graph) with Maven or Gradle modules without Java compiler warnings? That test case demonstrates the problem with Gradle, but it is the same problem with Maven. The `README.md` explains in details.
* If no, is it Maven's decision to not allow developers to build their project cleanly, at least regarding those standard Java warnings or errors?

Unless you can answer "yes" to one of above questions, multi-JAR files is a natural consequence of *one* way (among 2 or 3 ways) that Java standard compiler works today, and are **necessary** for allowing developers to build safer code and documentation in cases like above use case. Anyway, it is not in this thread that we can said "no, this safety doesn't deserve to be given to developers". This is a question for the community.

**Reply by rmannibucau** [reply]
The sample mixes topics.

1. Building  a mjar without warnings -> we already do since day 1 (think log4j does if it didnt change)
2. Module hierarchy (jre/jdk layout) is not supported by maven/gradle but is by mx or jdk toolkit, this is the opiniated choice of the build and maven does not this one. Agree it can be changed for v5 but it will also make it harder for all the mainstream cases where it is not relevant - keep in mind it was design for the jdk monolitic build

Anyway belongs to the list, anything discuss there has very few chance to be asf friendly

**Reply by desruisseaux** [reply]
> Building a mjar

If "mjar" means "multi-releases", the sample is not at all about this topic. It is only about Module Source Hierarchy.

> Module hierarchy (jre/jdk layout) _…snip…_ keep in mind it was design for the jdk monolitic build

Module Source Hierarchy is for any project that wish to use it. The OpenJDK project uses it, but there is nothing specific to OpenJDK. Not allowing developers to use it if they want seems more opinionated than giving this freedom.

**Reply by rmannibucau** [reply]
Ok, agree on mjar - thought at some point you want to merge the "same" module to do a mjar since it is/can be somehow related.

Still module source hierarchy was for the jdk itself originally.
While it can meet any project - and I'm thinking to Tomcat more than any other maven based projects - it must not go against the build tools and it does for maven (and gradle at some point).
If we stick to javac in the reasoning we'll miss the part that maven is a pipeline and the pipeline has assumptions, this does break it since 1 module = multiple main artifacts.

Side note agree with Gradle comment that it is probably not a good default case - indeed you can always enable it overriding default plugins: https://github.com/gradle/gradle/issues/25974#issuecomment-1836865066.

**Comment by desruisseaux** [comment]
> Side note agree with Gradle comment that it is probably not a good default case

That Gradle comment was based on a misunderstanding of what Module Source Hierarchy is. I'm in the Community Over Code Europe conference and had a talk with a Gradle engineer yesterday. They [reopened the issue](https://github.com/gradle/gradle/issues/25974#issuecomment-2144659457) following that discussion.

> it must not go against the build tools

It seems to me that Maven should adapt to Java, not the opposite.

**Reply by rmannibucau** [reply]
> It seems to me that Maven should adapt to Java, not the opposite.

You probably noted this is *not* the direction java takes, only the jdk, it is not even promoted by the oracle evangelists....

**Reply by desruisseaux** [reply]
There is no way it can be widely used as long as the tools do not support it. Note also that the Oracle evangelists do not neither said "don't use it". My guess is that they just acknowledge that they can't said "use it" if developers are unable to do so with Maven or Gradle.

**Reply by rmannibucau** [reply]
> There is no way it can be widely used as long as the tools do not support it.

I know it is your thing so let do it, please make it possible to create one folder per half of class - there is no user but until it is in maven there will be none ;).

>  My guess is that they just acknowledge that they can't said "use it" if developers are unable to do so with Maven or Gradle.

You said it was the same for JPMS but they promoted it a little bit emphasing the target was primarly the JDK itself so I guess this is not right.

Also note there is not a single use case we don't enable without that feature too so not sure that it is something we should focus on until we fixed all issues of the maven bugtracker - which can be saner than targetting new features for now.

**Comment by desruisseaux** [comment]
> You said it was the same for JPMS but they promoted it a little bit emphasing the target was primarly the JDK itself so I guess this is not right.

I remember to have seen this statement from an Oracle employee, but there is also material such as [Youtube videos](https://www.youtube.com/watch?v=M7q3C8OwJe8) of Oracle employees showing how to use JPMS in our own projects. There is also some new Java features that requires JPMS, such as [JEP-476](https://openjdk.org/jeps/476) (we way agree or not with that JEP, this is not the point). In addition of already existing tools that are clearly targeted to users, not the JDK, such as `jlink`.

**Reply by rmannibucau** [reply]
Does not change much that we dont need this thread feature to enable all jvm use cases...so noop is fine for me

**Reply by desruisseaux** [reply]
Noop may be fine for you, but not for other developers on this thread. Nobody is forcing you to use JPMS if you don't want. But conversely, please don't block those who want to use JPMS (and no, it is hardly possible with today's Maven. We tried very hard).

**Reply by rmannibucau** [reply]
I dont block anyone to do - even helped actually, I'm just trying to keep maven simple and not become ant as proposed in this thread for very few people.
